File: Introduction_to_Robotics_Course/Chapter9/Lesson3/Chapter9_Lesson3.m
function R = Rz(theta)
    c = cos(theta); s = sin(theta);
    R = [ c -s  0;
          s  c  0;
          0  0  1 ];
end

function xB = rigid_apply(R, p, xA)
    xB = R*xA + p;
end

function [R_ca, p_ca] = rigid_compose(R_cb, p_cb, R_ba, p_ba)
    R_ca = R_cb * R_ba;
    p_ca = R_cb * p_ba + p_cb;
end

% Demo
theta = pi/4;
R = Rz(theta);
p = [0.5; -0.2; 0.1];
xA = [1;0;0];
xB = rigid_apply(R, p, xA)

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter9/Lesson3/Chapter9_Lesson3.java
import org.ejml.simple.SimpleMatrix;

public class RigidMotion {
    public static SimpleMatrix Rz(double theta) {
        double c = Math.cos(theta), s = Math.sin(theta);
        double[][] data = {
            { c, -s, 0 },
            { s,  c, 0 },
            { 0,  0, 1 }
        };
        return new SimpleMatrix(data);
    }

    public static SimpleMatrix rigidApply(SimpleMatrix R, SimpleMatrix p, SimpleMatrix x) {
        return R.mult(x).plus(p);
    }

    public static SimpleMatrix[] rigidCompose(SimpleMatrix R_cb, SimpleMatrix p_cb,
                                              SimpleMatrix R_ba, SimpleMatrix p_ba) {
        SimpleMatrix R_ca = R_cb.mult(R_ba);
        SimpleMatrix p_ca = R_cb.mult(p_ba).plus(p_cb);
        return new SimpleMatrix[]{R_ca, p_ca};
    }

    public static void main(String[] args) {
        double theta = Math.PI/4.0;
        SimpleMatrix R = Rz(theta);
        SimpleMatrix p = new SimpleMatrix(3,1,true, new double[]{0.5, -0.2, 0.1});
        SimpleMatrix xA = new SimpleMatrix(3,1,true, new double[]{1,0,0});

        SimpleMatrix xB = rigidApply(R, p, xA);
        System.out.println("xB = " + xB);
    }
}
